jcapi-commands.go

```diff
@@ -0,0 +1,158 @@
+
+	Skip  int `json:"skip"`  // Objects to skip on /search POST
+	Limit int `json:"limit"` // Max objects to return on /search POST
+}
+
+func (e JCCommand) ToString() string {
+	return fmt.Sprintf("command: %v", e)
+}
+
+func getJCCommandsFromResults(result []byte) (commands []JCCommand, err JCError) {
+	commandResults := JCCommandResults{}
+
+	err = json.Unmarshal(result, &commandResults)
+	if err != nil {
+		err = fmt.Errorf("Could not unmarshal result '%s', err='%s'", string(result), err.Error())
+	}
+
+	commands = commandResults.Results
+
+	return
+}
+
+func (jc JCAPI) GetAllCommands() (commandList []JCCommand, err JCError) {
+	var empty []byte
+
+	for skip := 0; skip == 0 || len(commandList) == searchLimit; skip += searchSkipInterval {
+		url := fmt.Sprintf("%s?sort=hostname&skip=%d&limit=%d", COMMAND_PATH, skip, searchLimit)
+
+		jcSysRec, err2 := jc.DoBytes(MapJCOpToHTTP(Read), url, empty)
+
+		if err2 != nil {
+			return nil, fmt.Errorf("ERROR: Get commands to JumpCloud failed, err='%s'", err2)
+		}
+
+		if jcSysRec == nil {
+			return nil, fmt.Errorf("ERROR: No commands found")
+		}
+
+		resultsBlock, err2 := getJCCommandsFromResults(jcSysRec)
+		if err2 != nil {
+			err = fmt.Errorf("Could not get resultsBlock data, err='%s'", err2.Error())
+			return
+		}
+
+		for i, _ := range resultsBlock {
+			if resultsBlock[i].Id != "" {
+				commandList = append(commandList, resultsBlock[i])
+			}
+		}
+
+	}
+
+	return
+}
+
+func FindCommandById(commands []JCCommand, id string) (result *JCCommand, index int) {
+	index = FindObject(GetInterfaceArrayFromJCCommand(commands), "Id", id)
+	if index >= 0 {
+		result = &commands[index]
+	}
+
+	return
+}
+
+func GetInterfaceArrayFromJCCommand(commands []JCCommand) (interfaceArray []interface{}) {
+	interfaceArray = make([]interface{}, len(commands), len(commands))
+
+	for i := range commands {
+		interfaceArray[i] = commands[i]
+	}
+
+	return
+}
+
+//
+// Add or Update a command in place on JumpCloud
+//
+func (jc JCAPI) AddUpdateCommand(op JCOp, command JCCommand) (id string, err JCError) {
+	id, err = jc.HandleCommand(COMMAND_PATH, op, command)
+
+	return
+}
+
+func (jc JCAPI) HandleCommand(path string, op JCOp, command JCCommand) (id string, err JCError) {
+	data, err := json.Marshal(command)
+	if err != nil {
+		return "", fmt.Errorf("ERROR: Could not marshal JCCommand object, err='%s'", err.Error())
+	}
+
+	url := path
+	if op == Update {
+		url += "/" + command.Id
+	}
+
+	result, err := jc.DoBytes(MapJCOpToHTTP(op), url, data)
+	if err != nil {
+		return "", fmt.Errorf("ERROR: Could not '%s' new JCCommand object, err='%s'", MapJCOpToHTTP(op), err.Error())
+	}
+
+	commandResult := JCCommand{}
+
+	err = json.Unmarshal(result, &commandResult)
+	if err != nil {
+		return "", fmt.Errorf("ERROR: Could not unmarshal result '%s', err='%s'", string(result), err.Error())
+	}
+
+	return commandResult.Id, nil
+}
```

Comment:
This feels like a private worker function; do we want consumers to have access to this particular function, or just the more explicit crud actions?